changes/views.py

```diff
@@ -72,3 +78,54 @@ def post(self, request, *args, **kwargs):
                               action='momconnect_nonloss_optout',
                               data={'reason': 'sms_failure'})
         return Response(status=status.HTTP_201_CREATED)
+
+
+class ReceiveAdminOptout(mixins.CreateModelMixin,
+                         generics.GenericAPIView):
+    permission_classes = (IsAuthenticated,)
+
+    def post(self, request, *args, **kwargs):
+
+        try:
+            data = utils.json_decode(request.body)
+        except ValueError as e:
+            return JsonResponse({'reason': "JSON decode error",
+                                'details': six.text_type(e)}, status=400)
+
+        try:
+            identity_id = data['identity']
+        except KeyError as e:
+            return JsonResponse({'reason': '"identity" must be specified.'},
+                                status=400)
+
+        sbm_client = StageBasedMessagingApiClient(
+            api_url=settings.STAGE_BASED_MESSAGING_URL,
+            auth_token=settings.STAGE_BASED_MESSAGING_TOKEN
+        )
+
+        active_subs = sbm_client.get_subscriptions(
+            {'identity': identity_id, 'active': True}
+        )["results"]
+
+        actions = []
+        for sub in active_subs:
+            messageset = sbm_client.get_messageset(sub["messageset"])
+            if "nurseconnect" in messageset["short_name"]:
+                actions.append("nurse_optout")
+            elif "pmtct" in messageset["short_name"]:
+                actions.append("pmtct_nonloss_optout")
+            else:
+                actions.append("momconnect_nonloss_optout")
+
+        source = Source.objects.get(user=self.request.user)
+        request.data["source"] = source.id
+
+        for action in set(actions):
+            Change.objects.create(**{
+                "registrant_id": identity_id,
+                "action": action,
+                "data": {"reason": "other"},
+                "source": source,
+            })
+
+        return JsonResponse({})
```

Comment:
We probably want to return a `Response` instead of a `JsonReponse` here, so that Django Rest Framework can handle deciding if the response should be JSON or not, and handle encoding etc..

We should also return with an HTTP 201 result, and probably a list of the new change objects that were created